### PR TITLE
Add web_benchmarks package.

### PR DESCRIPTION
**Pull request by pennzht**
## Working in progress. Do not review yet.

This PR moves **Gallery web benchmarks** out of `flutter/flutter`. See https://github.com/flutter/flutter/pull/66351 for more information.

## References
web_benchmarks side — https://github.com/pennzht/web_benchmarks/pull/1
gallery side — https://github.com/flutter/gallery/pull/321
flutter side — https://github.com/flutter/flutter/pull/66351

### CODE REVIEW

File: lib/web_benchmarks.dart [hunk 1]
```diff
@@ -1,7 +1,5 @@
-library web_benchmarks;
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
Comment (yjbanov) [inline lib/web_benchmarks.dart]: Is this file needed?

File: lib/driver.dart [hunk 2]
```diff
@@ -0,0 +1,407 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
Comment (yjbanov) [inline lib/driver.dart:1]: There's a specific protocol between `driver.dart` and the local server. I'm afraid that it may cost more later to change this protocol because changing it requires changes in two repositories and there's no way to make the change in a single PR atomically. I see two options:

1. This package does not include a local server and only provides the `Recorder` API. The client/server protocol will continue living over in the flutter/flutter repo.
2. This package includes both the client and server implementation. The code in flutter/flutter would import both. The server API would not depend on/implement the "task" protocol used by the devicelab. Instead it would provide an API for collecting all the benchmark numbers. The devicelab task will be responsible for converting that to the devicelab protocol.

File: lib/driver.dart [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+import 'dart:async';
+import 'dart:convert' show json;
+import 'dart:html' as html;
+import 'dart:math' as math;
+
+import 'package:web_benchmarks/recorder.dart';
+
+typedef RecorderFactory = Recorder Function();
+
+const bool isCanvasKit = bool.fromEnvironment('FLUTTER_WEB_USE_SKIA', defaultValue: false);
```
Comment (yjbanov) [inline lib/driver.dart]: This is not needed here. Also, it depends on the internal `FLUTTER_WEB_USE_SKIA` flag which can change any time.

File: lib/driver.dart [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+
+  profile.scoreData.forEach((String scoreKey, Timeseries timeseries) {
+    html.document.body.appendHtml('<h2>$scoreKey</h2>');
+    html.document.body.appendHtml('<pre>${timeseries.computeStats()}</pre>');
+    html.document.body.append(TimeseriesVisualization(timeseries).render());
+  });
+}
+
+/// Draws timeseries data and statistics on a canvas.
+class TimeseriesVisualization {
```
Comment (yjbanov) [inline lib/driver.dart:141]: Any advantages to have the visualization in this package over keeping it in flutter/flutter?

File: lib/recorder.dart [hunk 5]
```diff
@@ -0,0 +1,1191 @@
...
+
+  _engineBenchmarkListeners[name] = listener;
+}
+
+/// Stops listening to engine benchmark values under labeled by [name].
+void stopListeningToEngineBenchmarkValues(String name) {
+  _engineBenchmarkListeners.remove(name);
+  if (_engineBenchmarkListeners.isEmpty) {
+    // The last listener unregistered. Remove the global listener.
+    js_util.setProperty(html.window, '_flutter_internal_on_benchmark', null);
```
Comment (yjbanov) [inline lib/recorder.dart:1179]: What would be the process for changing this logic when the internals of the engine need to change? Can this part move to the flutter/flutter repo so that this package is independent of engine-benchmark protocol?

File: test/server/framework/framework.dart [hunk 6]
```diff
@@ -0,0 +1,277 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
Comment (yjbanov) [inline test/server/framework/framework.dart]: I'm not sure this repo needs the task framework used by the devicelab.

File: test/server/framework/utils.dart [hunk 7]
```diff
@@ -0,0 +1,773 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
Comment (yjbanov) [inline test/server/framework/utils.dart]: How much of this file is actually used? (I guess this comment applies to any code copied out of devicelab)

**Review by yjbanov — COMMENTED** [review]
The overall approach should work. I have some tactical questions. One thing to be careful about is how we roll this framework relative to the various implicit protocols that it's implementing:

- The client/server protocol, where the client is supplied by this package but the server is supplied by flutter/flutter.
- The engine/benchmark protocol, where the engine side is supplied by the flutter/engine repo but the benchmark side is supplied by this package.

File: pubspec.yaml [hunk 8]
```diff
@@ -15,40 +15,11 @@ dependencies:
 dev_dependencies:
   flutter_test:
     sdk: flutter
+  args: ">=1.6.0"
+  logging: ">=0.11.4"
+  process: ">=3.0.13"
+  shelf: ">=0.7.5"
+  shelf_static: ">=0.2.8"
+  webkit_inspection_protocol: ">=0.7.3"
```
Comment (yjbanov) [inline pubspec.yaml:22]: It's great that all pub dependencies are dev_dependencies!

File: lib/driver.dart (hunk 2 above)
Comment (yjbanov) [reply lib/driver.dart:1]: It just occurred to me that option 2 has the disadvantage that this package would need more dependencies in order to implement the server (`package:shelf` and friends). Your current approach and option 2 don't have that problem.

File: lib/driver.dart (hunk 4 above)
Comment (pennzht) [reply lib/driver.dart:141]: This way we can run `flutter run -d chrome -t ... --profile` in the Gallery and see the results in the browser.

File: lib/recorder.dart (hunk 5 above)
Comment (pennzht) [reply lib/recorder.dart:1179]: This part is required by `WidgetRecorder`.

If we should move it to `flutter/flutter`, where should we move it to? It must be imported by `web_benchmarks`.

File: test/server/framework/framework.dart (hunk 6 above)
Comment (pennzht) [reply test/server/framework/framework.dart]: I removed unnecessary utils.

File: test/server/framework/utils.dart (hunk 7 above)
Comment (pennzht) [reply test/server/framework/utils.dart]: I removed unnecessary utils. I left a few that are used by `server.dart`.

File: lib/driver.dart (hunk 3 above)
Comment (pennzht) [reply lib/driver.dart]: Done.

File: lib/driver.dart (hunk 2 above)
Comment (pennzht) [reply lib/driver.dart:1]: I think it's good to keep it as is. It is definitely possible to reduce duplicate code by creating a common package for the server, but that seems unnecessary at this time.

File: lib/web_benchmarks.dart (hunk 1 above)
Comment (pennzht) [reply lib/web_benchmarks.dart]: It does seem unnecessary. Removed.

**Review by yjbanov — APPROVED** [review]
lgtm
